### ISSUE TITLE
Warn if context values passed to createPage are different types than others or empty

### ISSUE THREAD

**Issue by KyleAMathews** [gatsbyjs/gatsby#4360]
Someone creating pages could run into a mysterious GraphQL error if any of the pages context has missing information:

```
Errors:
GraphQLError: Variable "$slug" of required type "String!" was not provided.
```

This is hard for someone new to Gatsby to debug. We should check against this and warn.

**Comment by mainframev** [comment]
yeah, right now have this issue, trying to fix, but without result :(

**Comment by BerlinaLI** [comment]
i encounter this issue and found out it is because i didnot wirte slug in my gatsby node file
`
function createSchoolsPages (createPage, graphql) {
  const articleTemplate = path.resolve(`src/templates/partnered-schools/src/containers/School/School.js`)
  const listingTemplate = path.resolve(`src/templates/partnered-schools/src/containers/PartneredSchools/PartneredSchools.js`)

  return new Promise((resolve, reject) => {
    resolve(
      graphql(`
        query AllSchoolsQuery {
          allSourcePluginApiSchoolArticle {
            edges {
              node {
                slug
              }
            }
          }
        }

      `).then(result => {
        if (result.errors) {
          reject(result.errors)
        }
         })
    )
  })
}
`

**Comment by benjaminmodayil** [comment]
@Jonqth Thank you so much!

**Comment by peterjaap** [comment]
@Jonqth thanks so much for that last edit. Finally fixed this issue after hours of hunting..!

**Comment by alanshortis** [comment]
I have run into this error today - when creating pages from a folder of `.mdx` files, I was defining a component in `createPage` that lives in my `src/pages` folder. Moving my post component to `src/components` avoids it.

Any reason this should be the case? It's not a big deal, but I'd prefer to have the component in my pages folder since it's creating pages.

All pages are created as expected, I just get this error.

**Comment by tianzhich** [comment]
I have run into this error when migrating from v1 to v2, the context values passed to createPage are same types and not null. 

Anyone have same problems?

**Comment by whalesingswee** [comment]
@tianzhich I was able to fix the problem after moving my post to `/src/components` instead of using it inside `/src/pages` as @alanshortis  mentioned. Thanks!

**Comment by tianzhich** [comment]
I solved in #15977 cause the difference between StaticQuery and page query.

Maybe it could help someone.
